README
Table of Contents
=================

 1. Introduction
 2. Dependencies
 3. Building
 4. Bugs
 5. Translations
 6. Covers
 7. CUE Files
 8. Streams
 9. MultiMedia Keys
10. DBUS Actions
11. Dynamic Helper Script - Local Mode
12. Dynamic Helper Script - Server Mode
13. Source Code
14. Debug Logging
15. Credits
16. Windows
17. Mac OSX



1. Introduction
===============

Cantata is a graphical client for MPD. It contains the following features:

 1. Support for Linux, MacOSX, and Windows.
 2. Multiple MPD collections.
 3. Highly customisable layout.
 4. Songs grouped by album in play queue.
 5. Context view to show artist, album, and song information of current track.
 6. Simple tag editor.
 7. File organizer - use tags to organize files and folders.
 8. Ability to calculate ReplyGain tags. (Linux only, and if relevant libraries
    installed)
 9. Dynamic playlists.
10. Online services; Jamendo, Magnatune, SoundCloud, and Podcasts.
11. Radio stream support - with the ability to search for streams via TuneIn,
    ShoutCast, or Dirble.
12. USB-Mass-Storage and MTP device support. (Linux only, and if relevant
    libraries installed)
13. Audio CD ripping and playback. (Linux only, and if relevant libraries
    installed)
14. Playback of non-MPD songs - via simple in-built HTTP server if connected
    to MPD via a standard socket, otherwise filepath is sent to MPD.
15. MPRISv2 DBUS interface.
16. Basic support for touch-style interface (views are made 'flickable')
17. Scrobbling.
18. Ratings support.

Cantata started off as a fork of QtMPC, however, the code (and user interface)
is now *very* different to that of QtMPC.

Unlike most other MPD clients, Cantata caches a copy of the MPD music library.
This is so that it can create a proper hierarchy of artists and albums (where
Cantata will use the AlbumArtist tag if found, otherwise it will fallback to
the Artist tag), to provide album counts, tracks counts, etc in the browser
views and tooltips, and also to help with copying songs to/from devices.

This cache is updated only when MPD indicates that its version of the collection
is different to Cantata's.


2. Dependencies
===============

Cantata requires the following Qt libraries:

1. QtCore
2. QtGui
3. QtNetwork
4. QtXml
5. QtSql
6. QtDBus (Only for Linux builds)
7. QtWidgets (Qt5 builds - where widgets are now separated from QtGui)
8. QtMultimedia / libVLC - depends on ENABLE_HTTP_STREAM_PLAYBACK and
   ENABLE_LIBVLC
9. QtSvg

Cantata may also use the following optional libraries:

1. TagLib - Tag edit dialog, replaygain, track organizer, and UMS device
            support.
2. LibMTP - MTP devices.
3. FFMPEG (libavcodec) - ReplayGain detection code.
4. MPG123 - ReplayGain detection code.
5. CDParanoia - Read/rip Audio CDs.
6. CDDB - either this or MusicBrainz5 required for Audio CDs
7. MusicBrainz5 - either this or CDDB required for Audio CDs
8. LibEbur128 - ReplayGain detection code. If this is not found on the
   system then Cantata will use its own bundled copy.
9. libavahi-common / libavahi-client - Avahi support (automatic mpd discovery)
10. uchardet - automatic encoding detection


3. Building
===========

Please refer to the INSTALL file for general build instructions. Details on how
Cantata is build for Windows can be found in section 16, and Mac OSX in section
17.


4. Bugs
=======

Before reporting bugs, please check that this is not an MPD related issue with
your system - e.g. please try recreating the issue with another MPD client,
such as MPDroid, GMPC, etc.

To report bugs please visit https://github.com/CDrummond/cantata

When reporting a bug, please mention the Cantata version number, the Cantata
build type (Linux, Windows, Mac), any cmake options you specified
(other than the defaults listed at the top of the INSTALL file), and the
version of MPD you are using.


5. Translations
===============

As of version 2.1.0, Cantata uses Qt5's translation framework.

To generate a new translation for Cantata:

1. Copy translations/blank.ts to cantata_LANG_CODE.ts
   (Replace LANG_CODE with the language code for the desired language, e.g. 
   'fr' for French - so this would be cantata_fr.ts)
2. Open your new ts file with Qt Linguist.
3. In Qt Linguist set the target language and country.
4. Update the translations.
5. Either email me the new translation (address is in the AUTHORS file), or
   create a pull request on github.


6. Covers
=========

When displaying covers, Cantata will load album covers in the following order:

${configuredName} is either:
a. cover - if not configured in Cantata's config dialog
b. Setting from cantata with the following replacements:
    %albumartist% => Album Artist
    %artist% => Track Artist
    %album% => Album Name
    Any other % removed

...if configured to 'Cache scaled covers', Cantata will check its scaled-covers
   cache folder (~/.cache/cantata/covers-scaled/SIZE) for:

 1. ${albumArtist}/${album}.png

...if MPD folder exists, is not specified as a http URL, and is readable, then
   cantata will look for the following within the folder containing the song:

 2. ${configuredName}.jpg - if set
 3. ${configuredName}.png - if set
 4. cover.jpg
 5. cover.png
 6. AlbumArt.jpg
 7. AlbumArt.png
 8. folder.jpg
 9. folder.png
10. ${albumArtist} - ${album}.jpg
11. ${albumArtist} - ${album}.png
12. ${album}.jpg
13. ${album}.png
14. Image embedded within current song's tags.
15. ANY other jpg, or png

...then Cantata will check its cache folder (~/.cache/cantata/covers), for :

16. ${albumArtist}/${album}.jpg
17. ${albumArtist}/${album}.png

...if the MPD folder was specified as a http URL

18. ${url}/${dirFromFile}/cover.jpg
    - or ${url}/${dirFromFile}/${configuredName}.jpg (if set)
19. ${url}/${dirFromFile}/cover.png
    - or ${url}/${dirFromFile}/${configuredName}.png (if set)

...lastly, if user has enabled downloading via last.fm

20. Query last.fm using ${albumArtist} and ${album}. Cantata will attempt to
    download the image specified with the "extralarge" size.

Downloaded images will be saved as ${configuredName}.jpg/png (if set), or
cover.jpg/png, within the song folder if possible. If not, then they will be
saved in Cantata's cache folder.

If configured to 'Cache scaled covers', then the resized cover (as displayed)
will be saved as a PNG within covers-scaled.


For artist images:

${configuredName} is either:
a. artist - if configured Album cover ${configuredName} is empty, or does not
   contain %
b. ${albumArtist} - if configured Album cover ${configuredName} contains % and
   ${albumArtist} is not 'Various Artists'
c. ${basicArtist} if configured Album cover ${configuredName} contains % and
   ${albumArtist} is 'Various Artists'

...if configured to 'Cache scaled covers', Cantata will check its scaled-covers
   cache folder (~/.cache/cantata/covers-scaled/SIZE) for:

 1. ${albumArtist}.png

...if MPD folder exists, is not specified as a http URL, and is readable, then
   cantata will look for the following within the folder containing the song:

 2. ${basicArtist}.jpg
 3. ${basicArtist}.png
 4. ${configuredName}.jpg
 5. ${configuredName}.png

...the above will be repeated for the parent folder

...if song is from a Various Artists album, or file is a non-MPD file, then
   cantata will look for the following within the MPD root folder:

 6. ${basicArtist}/${basicArtist}.jpg
 7. ${basicArtist}/${basicArtist}.png
 8. ${basicArtist}/${configuredName}.jpg
 9. ${basicArtist}/${configuredName}.png

...then Cantata will check its cache folder (~/.cache/cantata/covers), for :

10. ${albumArtist}.jpg
11. ${albumArtist}.png

...if the MPD folder was specified as a http URL

12. ${url}/${dirFromFile}/${configuredName}.jpg
13. ${url}/${dirFromFile}/${configuredName}.png
 
...the above will be repeated for the parent folder

...lastly, if user has enabled downloading via last.fm

14. Query last.fm using ${albumArtist}. Cantata will attempt to download the
    image specified with the "extralarge" size.

Downloaded images will be saved as artist.jpg/png, within the artist folder
if the folder hierarchy is 2 levels (artist/album/) and the user has write
permissions. If not, then they will be saved in Cantata's cache folder.

If configured to 'Cache scaled covers', then the resized cover (as displayed) 
will be saved as a PNG within covers-scaled.


For composer images:

...if configured to 'Cache scaled covers', Cantata will check its scaled-covers
   cache folder (~/.cache/cantata/covers-scaled/SIZE) for:

1. ${composer}.png

...if MPD folder exists, is not specified as a http URL, and is readable, then
   cantata will look for the following within the folder containing the song:

2. composer.jpg
3. composer.png

...the above will be repeated for the parent folder

...then Cantata will check its cache folder (~/.cache/cantata/covers), for :

4. ${composer}.jpg
5. ${composer}.png

...if the MPD folder was specified as a http URL

6. ${url}/${dirFromFile}/composer.jpg
7. ${url}/${dirFromFile}/composer.png

...the above will be repeated for the parent folder

Downloaded images will be saved as composer.jpg/png, within the artist folder
if the folder hierarchy is 2 levels (artist/album/) and the user has write
permissions. If not, then they will be saved in Cantata's cache folder.

If configured to 'Cache scaled covers', then the resized cover (as displayed)
will be saved as a PNG within covers-scaled.


For context view backdrops:

...if MPD folder exists, is not specified as a http URL, and is readable, OR
   the song's filename starts with '/', then cantata will look for the
   following within the folder containing the song:

 1. ${currentArtist}-backdrop.jpg
 2. ${currentArtist}-backdrop.png
 3. backdrop.jpg
 4. backdrop.png

...the above will be repeated for the parent folder

...if song is from a Various Artists album, or file is a non-MPD file, then
   cantata will look for the following within the MPD root folder:

 5. ${currentArtist}/${currentArtist}-backdrop.jpg
 6. ${currentArtist}/${currentArtist}-backdrop.png
 7. ${currentArtist}/backdrop.jpg
 8. ${currentArtist}/backdrop.png

...then Cantata will check its cache folder (~/.cache/cantata/backdrops), for :

 9. ${currentArtist}.jpg

...internet services:

10. MusizBrainz is queried for an artist ID. If returned, this artist ID is used
    to locate a cover from fanart.tv
11. Download image from discogs

...lastly:

12. If all the above fails, a backdrop is created from all the album covers
    featuring the artist.

Downloaded images will be saved as backdrop.jpg, within the artist folder
if the current song is not from a Various Artists album, the file is from MPD,
the folder hierarchy is 2 levels (artist/album/), and the user has write
permissions. If not, then they will be saved in Cantata's cache folder.


7. CUE Files
============

If Cantata is running with MPD>=0.17, then it will attempt to parse the contents
of CUE files. The CUE file will be parsed to ascertain the album and track
details. The track durations for all bar the last should be accurate, and
Cantata will attempt to determine the duration of the last track - but this
might not be 100% accurate.

If the CUE file references an audio file that does not exist in the MPD music
folder, then the CUE file contents will NOT be used.

There is no reliable way for Cantata to ascertain the text encoding that a CUE
file uses - so, for portability, it would be best if your CUE files used UTF-8
encoding. When loading a CUE file, Cantata will attempt to load the file as if
it was UTF-8 encoded, if this fails it will try with the "System" codec. You may
add to this list of encodings to try by using the 'cueFileCodecs' config item
(as detailed in section 7 above). If Cantata fails with all configured
encodings, then it will 'peek' at the 1st 1k bytes, and ask Qt to see if it can
determine the encoding - and fallback to UTF-8 otherwise. (This peeking and
fallback was how Cantata behaved in pre 1.2 versions - and is how Clementine
behaves (where Cantata's CUE file support originates from.))

NOTE: If Cantata cannot access the CUE file (as MPD is running on a remote host,
or you have not configured the music folder correctly, etc.) then Cantata will
add the CUE File to the album's track listing.


8. Streams
==========

As of 1.4.0, Cantata only comes with TuneIn, ShoutCast, IceCast and Dirble
stream providers. To install new providers, you can either use Cantata's
download dialog (to install for Cantata's extras github repo), or install from
file. To import new stream categories into Cantata, these files need to be GZip
compressed tarballs of the following format.

  Filename: CATEGORY_NAME.streams  (e.g. German Radio.streams)

  Contents: Either streams.xml, streams.xml.gz, OR settings.json, AND
            Either icon.png OR icon.svg

            streams.xml.gz should be GZip compressed.
            icon.svg should be a non-compressed SVG file.

            Icons for sub-categories (first level only) - these should be named
            the same as the category (with / replaced by _) -
            e.g. 'Hard_Soft Rock.svg' for 'Hard/Soft Rock')

settings.json is really only intended to be used by Cantata for providers
stored in its extras github repo. This JSON file has the following syntax:

  { "type": "streamType", "url": "Stream listing URL" }
  
  "streamType" can be either di, soma, or listenlive. Cantata contains code to
  download from "Stream listing URL" and parse the response based upon
  "streamType"

In general, for externally provided stream lists, it is more convenient to just
list all a providers streams within stream.xml.gz. This has the following
format:

  <?xml version="1.0" encoding="UTF-8"?>
  <streams version="1.0" addCategoryName="true">
    <stream name="Station Name" url="http://station.stream.url"/>
  </streams>

if "addCategoryName" is set to true, then when Cantata adds a station from this
category to either favourites or the playqueue, it will prepend the Category
name to the station name.

e.g. To create a stream category named "Wicked Radio" that has 2
sub-categeries, and 3 stream URLs you would need:

streams.xml.gz with:
  <?xml version="1.0" encoding="UTF-8"?>
  <streams version="1.0" addCategoryName="true">
    <category name="Rock">
      <stream name="Hard Rock" url="http://127.0.0.1/hard.ogg"/>
      <stream name="Soft Rock" url="http://127.0.0.1/soft.ogg"/>
    </category>
    <stream name="Blues" url="http://127.0.0.1/blues.ogg"/>
  </streams>

streams.xml.gz & icon.svg need to be placed into a GZip compressed tarball
named "Wicked Radio.streams"

With the above example, Cantata would list the following in the streams view:

> Wicked Radio
   > Rock
       Hard Rock
       Soft Rock
   Blues

When "Hard Rock" is added to the playqueue, it will be listed as
"Wicked Radio - Hard Rock"

If Rock.png (or Rock.svg) and/or Blues.png (or Blues.svg) also exist in the
tarball, then they will be copied to the install location and used as icons for
the respective category.

To create .streams file:
    1. Place all files in a temp folder
    2. cd into temp folder
    3. tar cfz NAME.tar.gz *
    4. mv NAME.tar.gz NAME.streams


cantata-extra on github
-----------------------

The streams page on Cantata's settings dialog, will allow you to install new
providers from Cantata's extras github repo. To do this, the dialog first
downloads the list of providers, this is obtained from the following URL:

https://raw.githubusercontent.com/CDrummond/cantata-extra/master/streams/2.1/list.xml

The contents of this file will be something like:

<providers>
 <category type="0">
  <provider name="1.fm" md5="cbd6ad834d4493662952a653cea5c779"/>
  ...

If you want to manually download a provider, use the listed URL to download the
.streams file. e.g. for the above:

    wget https://raw.githubusercontent.com/CDrummond/cantata-extra/master/streams/2.1/1.fm.streams.gz

Extract the contents with:

    tar zxvf 1.fm.streams.gz

If you wish to distribute a new stream provider via Cantata's extras github
repo, please send me an email with the .streams.gz file. My address is in the AUTHORS
file.


9. MultiMedia Keys
==================

Linux and Windows builds, as of 1.2, now also support basic media-keys
shortcuts.


ChangeLog
2.5.x
-----
 1. uchardet detection integrated, this fixes library view with "???.."
 2. CUE playlist parser fix - FILE section now working correctly.
 3. Audio source file existance check from FILE section of CUE file.
 4. CUE parser optimizations.

2.5.0
-----
 1. Update translations.
 2. Limit number of album tracks shown in context view to 500, thanks to
    ccoors.
 3. Fix Community Radio Browser search.
 4. Remove dirble from radio section, as its no longer active.
 5. Better handling of CUE tracks when MPD is set to list as directory.
 6. Disable CUE parsing in cantata by default, as MPD handles this better now.
 7. Remember, and restore, main window position.
 8. Disable categorized view, as its been reported to crash (#1530)
 9. Remove stream providers, as many broken.
10. Fix decoding URLs when playing local files via in-built HTTP server.
11. Remove option to select cover image providers, always use all.
12. Remove Google and Spotify image search, not working.
13. Allow smaller images in itemviews.
14. Fix newlines showing as HTML tags in contextview.
15. Fix updating now-plying metadata for radio streams that transmit track
    numbers.
16. When stopping Cantata controlled MPD instance, wait up to 2 seconds for
    MPD to gracefully terminate (so config can be saved) before killing
    process.
17. Add support for MPD's "Partitions" - implemented by dphoyes. Requires MPD
    0.22 or above.
18. Allow queue to be sorted by path.
19. Fix some deprecation warnings - thanks to John Regan.
20. Fix crash when trying to copy songs to MTP device but libMTP has failed to
    get storage list.
21. Don't save queue if string entered in dave dialog but cancel button
    pressed.
22. Handle case where IceCast list is not GZipped.
23. Remove SoundCloud support, no longer works due to API changes.
24. Correctly update play queue time when re-order tracks - thanks to Philip
    Sequeira.
25. When searching for lyrics, if fail and artist starts with "The " then try
    again without "The "
26. Add "Refresh" action to hover actions for podcasts.
27. Remove superfluous blank space from the top of the cover tooltip.
28. Fix looking for cover-art with MPD's new cue track file listing.
29. Add Grouping tag support to playlists and play queue.
30. Use QCollator to compare strings.
31. If using table-style play queue, then only sort one column at a time.
32. Stop user MPD instance from GUI thread when terminating, to ensure state
    can be saved.
33. Don't write empty genres to tags.

2.4.2
-----
1. Correctly handle changing 'Basic' mode music folder.
2. When stopping 'Basic' mode MPD instance, send SIGKILL.
3. Correctly set 'storeLyricsInMpdDir' config item, UI was setting wrong config
   item.
4. Set minimum Qt5 version to 5.11
5. Fix destructor of DeviceManagerPrivate to prevent Cantata from potentially
   crashing when closing.
6. Correctly set song details 'time', 'year', 'track' and 'disc' for streams
   from online services.
7. Don't use last.fm for artist image searches, its broken.
8. Hide BB10 styles (look bad), and gtk2 style (doesn't start) from list of
   styles in interface settings.
9. When checking if song exists, check disc number.
10. Fix getting song details from Cantata stream URLs.
11. Amend MPRIS interface: fix CanPlay/CanPause/CanSeek status update as well
    as LoopStatus getter and setter.
12. When checking if songs are different, compare track and disc numbers too.
13. Avoid unknown song durations and duplicate updates of MPRIS' song details.
14. Synchronize update of current song details and status of MPRIS interface,
    trigger MPRIS status update when connection to MPD has been lost.
15. Add chartlyrics.com to list of lyrics providers.
16. Set default lyrics providers to azlyrics.com, chartlyrics.com, and
    lyrics.wikia.com
17. Fix enabling of play queue navigation actions 'next' and 'previous'.
18. Fix bus name of freedesktop.org's power management.
19. Additionally call Inhibit() from org.freedesktop.login1.Manager.
20. Query Qt whether system tray is available if current desktop environment is
    not some kind of GNOME (incl. Unity flavored GNOME).
21. Fix writing 'descr' attribute when saving podcast information to cache dir.
22. Fix loading cover images with wrong file extension in context view.
23. Avoid prepending song's file path with MPD's music directory if it is empty,
    a stream URL or an absolute path.
24. Ignore current song in selection when moving selected songs within the play
    queue to play them next.
25. Also show metadata of the current track in the context view if 'title' or
    'artist' are missing, but do not try to fetch information nor lyrics.
26. Switch from freedb.org to gnudb.org
27. Update bitrate settings for encoders used with transcoding jobs.

2.4.1
-----
1. Re-enable custom playqueue background. This is broken for 5.12, but that's a
   Qt bug.
2. Look in /usr/lib64/qt5/bin for lrelease
3. Fix deprecation warnings.
4. Enable catagorized view by default, might also be a Qt issue?
5. Remove Encyclopaedia Metallum from lyrics providers, as does not work.

2.4.0
-----
 1. Add 'Read offset' setting for AudioCDs.
 2. Show invalid files in playlists using red text.
 3. Add 'Remove Invalid Tracks' to playlist context menu.
 4. Allow OPML URLs in podcast add URL dialog.
 5. Allow to read local RSS/OPML files in podcast search dialog.
 6. Add action to export current podcast subscriptions to OPML file.
 7. Add searching for radio stations on Community Radio Browser.
 8. Show bits in technical info.
 9. Fix saving, and loading, of custom API keys.
10. Fix saving, and reading back, https:// as MPD music folder.
11. Fix crash when double-clicking outside of table-view (when this is set to
    not stretch columns).
12. Fix greyscale images in notifications.
13. Re-add option to save lyrics in music folder.
14. Show Original Year in context view metadata.
15. Add --fullscreen command-line option to start fullscreen.

online/jamendoservice.cpp
/*
 * Cantata
 *
 * Copyright (c) 2017-2022 Craig Drummond <craig.p.drummond@gmail.com>
 *
 * ----
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

#include "jamendoservice.h"
#include "jamendosettingsdialog.h"
#include "db/onlinedb.h"
#include "models/roles.h"
#include "support/icon.h"
#include "support/configuration.h"
#include "support/utils.h"
#include "support/monoicon.h"
#include <QXmlStreamReader>
#include <QUrl>

#ifdef TAGLIB_FOUND

#ifdef UCHARDET_FOUND
#include "support/uchardet.h"
#endif

#include <taglib/tstring.h>
#include <taglib/id3v1genres.h>
#include <QTextCodec>
static QString id3Genre(int id)
{
    QTextCodec *codec;
    #ifdef UCHARDET_FOUND
    static const char *uchardetCodec = UCHARDET::detectedEncoding(TagLib::ID3v1::genre(id).toCString(true));
    if (uchardetCodec != NULL) {
        codec = QTextCodec::codecForName(uchardetCodec);
    } else {
    #endif
    codec = QTextCodec::codecForName("UTF-8");
    #ifdef UCHARDET_FOUND
    }
    #endif
    // Clementine: In theory, genre 0 is "blues"; in practice it's invalid.
    return 0==id ? QString() : codec->toUnicode(TagLib::ID3v1::genre(id).toCString(true)).trimmed();
}

#else // TAGLIB_FOUND
static QString id3Genre(int id)
{
    // Clementine: In theory, genre 0 is "blues"; in practice it's invalid.
    if (0==id) {
        return QString();
    }

    static QMap<int, QString> idMap;

    if (idMap.isEmpty()) {
        idMap.insert(0, "Blues");
        idMap.insert(1, "Classic Rock");
        idMap.insert(2, "Country");
        idMap.insert(3, "Dance");
        idMap.insert(4, "Disco");
        idMap.insert(5, "Funk");
        idMap.insert(6, "Grunge");
        idMap.insert(7, "Hip-Hop");
        idMap.insert(8, "Jazz");
        idMap.insert(9, "Metal");
        idMap.insert(10, "New Age");
        idMap.insert(11, "Oldies");
        idMap.insert(12, "Other");
        idMap.insert(13, "Pop");
        idMap.insert(14, "R&B");
        idMap.insert(15, "Rap");
        idMap.insert(16, "Reggae");
        idMap.insert(17, "Rock");
        idMap.insert(18, "Techno");
        idMap.insert(19, "Industrial");
        idMap.insert(20, "Alternative");
        idMap.insert(21, "Ska");
        idMap.insert(22, "Death Metal");
        idMap.insert(23, "Pranks");
        idMap.insert(24, "Soundtrack");
        idMap.insert(25, "Euro-Techno");
        idMap.insert(26, "Ambient");
        idMap.insert(27, "Trip-Hop");
        idMap.insert(28, "Vocal");
        idMap.insert(29, "Jazz+Funk");
        idMap.insert(30, "Fusion");
        idMap.insert(31, "Trance");
        idMap.insert(32, "Classical");
        idMap.insert(33, "Instrumental");
        idMap.insert(34, "Acid");
        idMap.insert(35, "House");
        idMap.insert(36, "Game");
        idMap.insert(37, "Sound Clip");
        idMap.insert(38, "Gospel");
        idMap.insert(39, "Noise");
        idMap.insert(40, "AlternRock");
        idMap.insert(41, "Bass");
        idMap.insert(42, "Soul");
        idMap.insert(43, "Punk");
        idMap.insert(44, "Space");
        idMap.insert(45, "Meditative");
        idMap.insert(46, "Instrumental Pop");
        idMap.insert(47, "Instrumental Rock");
        idMap.insert(48, "Ethnic");
        idMap.insert(49, "Gothic");
        idMap.insert(50, "Darkwave");
        idMap.insert(51, "Techno-Industrial");
        idMap.insert(52, "Electronic");
        idMap.insert(53, "Pop-Folk");
        idMap.insert(54, "Eurodance");
        idMap.insert(55, "Dream");
        idMap.insert(56, "Southern Rock");
        idMap.insert(57, "Comedy");
        idMap.insert(58, "Cult");
        idMap.insert(59, "Gangsta");
        idMap.insert(60, "Top 40");
        idMap.insert(61, "Christian Rap");
        idMap.insert(62, "Pop/Funk");
        idMap.insert(63, "Jungle");
        idMap.insert(64, "Native American");
        idMap.insert(65, "Cabaret");
        idMap.insert(66, "New Wave");
        idMap.insert(67, "Psychadelic");
        idMap.insert(68, "Rave");
        idMap.insert(69, "Showtunes");
        idMap.insert(70, "Trailer");
        idMap.insert(71, "Lo-Fi");
        idMap.insert(72, "Tribal");
        idMap.insert(73, "Acid Punk");
        idMap.insert(74, "Acid Jazz");
        idMap.insert(75, "Polka");
        idMap.insert(76, "Retro");
        idMap.insert(77, "Musical");
        idMap.insert(78, "Rock & Roll");
        idMap.insert(79, "Hard Rock");
        idMap.insert(80, "Folk");
        idMap.insert(81, "Folk-Rock");
        idMap.insert(82, "National Folk");
        idMap.insert(83, "Swing");
        idMap.insert(84, "Fast Fusion");
        idMap.insert(85, "Bebob");
        idMap.insert(86, "Latin");
        idMap.insert(87, "Revival");
        idMap.insert(88, "Celtic");
        idMap.insert(89, "Bluegrass");
        idMap.insert(90, "Avantgarde");
        idMap.insert(91, "Gothic Rock");
        idMap.insert(92, "Progressive Rock");
        idMap.insert(93, "Psychedelic Rock");
        idMap.insert(94, "Symphonic Rock");
        idMap.insert(95, "Slow Rock");
        idMap.insert(96, "Big Band");
        idMap.insert(97, "Chorus");
        idMap.insert(98, "Easy Listening");
        idMap.insert(99, "Acoustic");
        idMap.insert(100, "Humour");
        idMap.insert(101, "Speech");
        idMap.insert(102, "Chanson");
        idMap.insert(103, "Opera");
        idMap.insert(104, "Chamber Music");
        idMap.insert(105, "Sonata");
        idMap.insert(106, "Symphony");
        idMap.insert(107, "Booty Bass");
        idMap.insert(108, "Primus");
        idMap.insert(109, "Porn Groove");
        idMap.insert(110, "Satire");
        idMap.insert(111, "Slow Jam");
        idMap.insert(112, "Club");
        idMap.insert(113, "Tango");
        idMap.insert(114, "Samba");
        idMap.insert(115, "Folklore");
        idMap.insert(116, "Ballad");
        idMap.insert(117, "Power Ballad");
        idMap.insert(118, "Rhythmic Soul");
        idMap.insert(119, "Freestyle");
        idMap.insert(120, "Duet");
        idMap.insert(121, "Punk Rock");
        idMap.insert(122, "Drum Solo");
        idMap.insert(123, "A capella");
        idMap.insert(124, "Euro-House");
        idMap.insert(125, "Dance Hall");
    }

    return idMap[id];
}
#endif

static const QLatin1String constStreamUrl("http://api.jamendo.com/get2/stream/track/redirect/?id=%1&streamencoding=");
static const QLatin1String constListingUrl("http://imgjam.com/data/dbdump_artistalbumtrack.xml.gz");
static const QLatin1String constName("jamendo");

int JamendoXmlParser::parse(QXmlStreamReader &xml)
{
    int artistCount=0;
    QList<Song> *songList=new QList<Song>();
    while (!xml.atEnd()) {
        xml.readNext();
        if (QXmlStreamReader::StartElement==xml.tokenType() && QLatin1String("artist")==xml.name()) {
            parseArtist(songList, xml);
            artistCount++;
            if (songList->count()>500) {
                emit songs(songList);
                songList=new QList<Song>();
            }
        }
    }
    if (songList->isEmpty()) {
        delete songList;
    } else {
        emit songs(songList);
    }
    return artistCount;
}

void JamendoXmlParser::parseArtist(QList<Song> *songList, QXmlStreamReader &xml)
{
    Song song;

    while (!xml.atEnd()) {
        xml.readNext();

        if (QXmlStreamReader::StartElement==xml.tokenType()) {
            QStringRef name = xml.name();

            if (QLatin1String("name")==name) {
                song.artist=xml.readElementText().trimmed();
            } else if (QLatin1String("album")==name) {
                parseAlbum(song, songList, xml);
            } /*else if (artist && QLatin1String("image")==name) {
                artist->setImageUrl(xml.readElementText().trimmed());
            }*/
        } else if (xml.isEndElement() && QLatin1String("artist")==xml.name()) {
            break;
        }
    }
}

void JamendoXmlParser::parseAlbum(Song &song, QList<Song> *songList, QXmlStreamReader &xml)
{
    QString id;
    QString genre;
    song.track=0;
    song.album=QString();

    while (!xml.atEnd()) {
        xml.readNext();

        if (QXmlStreamReader::StartElement==xml.tokenType()) {
            QStringRef name = xml.name();

            if (QLatin1String("name")==name) {
                song.album=xml.readElementText().trimmed();
            } else if (QLatin1String("track")==name) {
                song.track++;
                parseSong(song, genre, xml);
                songList->append(song);
            } else if (QLatin1String("id")==name) {
                id=xml.readElementText().trimmed();
            } else if (QLatin1String("id3genre")==name) {
                int g=xml.readElementText().toInt();
                if (0!=g) {
                    genre=id3Genre(g);
                }
            }
        } else if (xml.isEndElement() && QLatin1String("album")==xml.name()) {
            break;
        }
    }

    if (!id.isEmpty()) {
        emit coverUrl(song.albumArtistOrComposer(), song.album, id);
    }
}

void JamendoXmlParser::parseSong(Song &song, const QString &albumGenre, QXmlStreamReader &xml)
{
    song.time=0;
    song.title=QString();
    song.genres[0]=albumGenre;

    while (!xml.atEnd()) {
        xml.readNext();

        if (QXmlStreamReader::StartElement==xml.tokenType()) {
            QStringRef name = xml.name();

            if (QLatin1String("name")==name) {
                song.title=xml.readElementText().trimmed();
            } else if (QLatin1String("duration")==name) {
                song.time=xml.readElementText().toFloat();
            } else if (QLatin1String("id3genre")==name && albumGenre.isEmpty()) {
                int g=xml.readElementText().toInt();
                if (0!=g) {
                    song.genres[0]=id3Genre(g);
                }
            } else if (QLatin1String("id")==name) {
                song.file=xml.readElementText().trimmed();
            }
        } else if (xml.isEndElement() && QLatin1String("track")==xml.name()) {
            break;
        }
    }
    song.fillEmptyFields();
}

static QString formatStr(JamendoService::Format f)
{
    return JamendoService::FMT_MP3==f ? "mp3" : "ogg";
}

static JamendoService::Format toFormat(const QString &f)
{
    return f=="ogg" ? JamendoService::FMT_Ogg : JamendoService::FMT_MP3;
}

JamendoService::JamendoService(QObject *p)
    : OnlineDbService(new OnlineDb(constName, p), p)
{
    icn=MonoIcon::icon(FontAwesome::playcircleo, Utils::monoIconColor());
    useCovers(name());
    Configuration cfg(constName);
    format=toFormat(cfg.get("format", formatStr(FMT_MP3)));
}

QVariant JamendoService::data(const QModelIndex &index, int role) const
{
    if (index.isValid()) {
        switch (role) {
        case Cantata::Role_CoverSong: {
            QVariant v;
            Item *item = static_cast<Item *>(index.internalPointer());
            switch (item->getType()) {
            case T_Album:
                if (item->getSong().isEmpty()) {
                    Song song;
                    song.artist=item->getParent()->getId();
                    song.album=item->getId();
                    song.setIsFromOnlineService(constName);
                    song.file=constName; // Just so that isEmpty() is false!
                    QString id=static_cast<OnlineDb *>(db)->getCoverUrl(/*T_Album==topLevel() ? static_cast<AlbumItem *>(item)->getArtistId() : */item->getParent()->getId(), item->getId());
                    song.setExtraField(Song::OnlineImageUrl, QString("http://api.jamendo.com/get2/image/album/redirect/?id=%1&imagesize=300").arg(id));
                    item->setSong(song);
                }
                v.setValue<Song>(item->getSong());
                break;
            case T_Artist:
                break;
            default:
                break;
            }
            return v;
        }
        }
    }
    return OnlineDbService::data(index, role);
}

QString JamendoService::name() const
{
    return constName;
}

QString JamendoService::title() const
{
    return QLatin1String("Jamendo");
}

QString JamendoService::descr() const
{
    return tr("The world's largest digital service for free music");
}

OnlineXmlParser * JamendoService::createParser()
{
    return new JamendoXmlParser();
}

QUrl JamendoService::listingUrl() const
{
    return QUrl(constListingUrl);
}

Song & JamendoService::fixPath(Song &s) const
{
    s.file=QString(constStreamUrl).replace("id=%1", "id="+s.file);
    s.file+=FMT_MP3==format ? QLatin1String("mp31") : QLatin1String("ogg2");
    s.type=Song::OnlineSvrTrack;
    s.setIsFromOnlineService(name());
    return encode(s);
}

void JamendoService::configure(QWidget *p)
{
    JamendoSettingsDialog dlg(p);
    if (dlg.run(FMT_MP3==format)) {
        Format f=0==dlg.format() ? FMT_MP3 : FMT_Ogg;
        if (f!=format) {
            format=f;
            Configuration cfg(constName);
            cfg.set("format", formatStr(format));
        }
    }
}

#include "moc_jamendoservice.cpp"
